projects/lib/src/board/seirawanboard.cpp

```diff
@@ -0,0 +1,440 @@
+	int baserank = (side == Side::White) ? 0 : height() -1 ;
+	return  rank == baserank && !side.isNull();
+}
+
+bool SeirawanBoard::variantHasOptionalPromotions() const
+{
+	return true;
+}
+
+QList<Piece> SeirawanBoard::reservePieceTypes() const
+{
+	QList<Piece> list;
+
+	for (int i = 0; i < 2; i++)
+	{
+		list << Piece(Side::Type(i), Hawk);
+		list << Piece(Side::Type(i), Elephant);
+	}
+
+	return list;
+}
+
+void SeirawanBoard::addPromotions(int sourceSquare,
+				  int targetSquare,
+				  QVarLengthArray<Move>& moves) const
+{
+	WesternBoard::addPromotions(sourceSquare, targetSquare, moves);
+
+	moves.append(Move(sourceSquare, targetSquare, Hawk));
+	moves.append(Move(sourceSquare, targetSquare, Elephant));
+}
+
+bool SeirawanBoard::vSetFenString(const QStringList& fen)
+{
+	m_squareMap.clear();
+	return WesternBoard::vSetFenString(fen);
+}
+
+void SeirawanBoard::insertIntoSquareMap(int square, int count)
+{
+	m_squareMap.insert(square, count);
+}
+
+void SeirawanBoard::updateSquareMap(const Move& move, int increment)
+{
+	int source = move.sourceSquare();
+	if (m_squareMap.contains(source))
+		m_squareMap[source] += increment;
+
+	int target = move.targetSquare();
+	if (m_squareMap.contains(target))
+		m_squareMap[target] += increment;
+}
+
+bool SeirawanBoard::parseCastlingRights(QChar c)
+{
+	Side side = (c.isUpper()) ? upperCaseSide() : upperCaseSide().opposite();
+	QChar ch = c.toLower();
+	QString rank = side == Side::White ? "1" : QString::number(height());
+
+	if (ch >= 'a' && ch <= 'h')
+	{
+		insertIntoSquareMap(squareIndex(ch + rank));
+		return true;
+	}
+	if (ch == 'k')
+	{
+		insertIntoSquareMap(kingSquare(side));
+		insertIntoSquareMap(squareIndex("h" + rank));
+	}
+	if (ch == 'q')
+	{
+		insertIntoSquareMap(kingSquare(side));
+		insertIntoSquareMap(squareIndex("a" + rank));
+	}
+	return WesternBoard::parseCastlingRights(c);
+}
+
+QString SeirawanBoard::vFenString(Board::FenNotation notation) const
+{
+	QString vfs = Chess::WesternBoard::vFenString(notation);
+	int reservePieceCount[2] = {0, 0};
+
+	const QList<Piece>& reserveTypes = reservePieceTypes();
+	for (const Piece& pc: reserveTypes)
+		reservePieceCount[pc.side()] += reserveCount(pc);
+
+	// prepend castling field with list of files usable for piece channeling
+	QString s;
+
+	for (int i: m_squareMap.keys())
+	{
+		if (m_squareMap[i] > 0)
+			continue;
+
+		Piece piece = pieceAt(i);
+		Side side = piece.side();
+		if (!variantHasChanneling(side, i))
+		{
+			qWarning() << "channeling map mismatch for square "
+				   << i << ": " << squareString(i);
+			continue;
+		}
+		// do not list channeling files if there are no reserve pieces
+		if (reservePieceCount[side] <= 0)
+			continue;
+
+		if (side == upperCaseSide())
+			s.append(squareString(i).mid(0,1).toUpper());
+		else
+			s.append(squareString(i).mid(0,1).toLower());
+	}
+
+	QStringList vfen = vfs.split(' ', QString::SkipEmptyParts);
+
+	if ((vfen.count() > 0 && vfen.at(0)!= "-") || s.length() == 0)
```

Comment:
In Qt the convention is to use `QStringList::isEmpty()` or `QString::isEmpty()` to test for emptiness.

Reply:
ah, thank you, will be changed